File: JobTracker/Services/NIMClient.swift
//
//  NIMClient.swift
//  JobTracker
//
//  Client for NVIDIA NIM's OpenAI-compatible chat completions endpoint.
//  The API key is read from the Keychain at call time.
//

import Foundation

enum NIMError: LocalizedError {
    case missingAPIKey
    case http(Int, String)
    case rateLimited
    case emptyResponse

    var errorDescription: String? {
        switch self {
        case .missingAPIKey:     return "No NVIDIA API key set. Add it in Settings."
        case .http(let c, let b): return "NVIDIA API HTTP \(c): \(b)"
        case .rateLimited:       return "NVIDIA API rate limit hit; will retry."
        case .emptyResponse:     return "NVIDIA API returned no content."
        }
    }
}

/// Paces all LLM requests to stay under the provider's rate limit
/// (NVIDIA free tier: 40 requests/minute → one request per 1.5 s; a small
/// margin keeps bursts safely below the cap). Every request in the app
/// funnels through NIMClient, so this is the single enforcement point.
@MainActor
final class RequestThrottle {
    static let shared = RequestThrottle()
    private var lastRequestStart: Date = .distantPast
    private let minInterval: TimeInterval = 1.6

    private init() {}

    /// Waits until it's safe to start the next request.
    func waitTurn() async {
        let elapsed = Date().timeIntervalSince(lastRequestStart)
        if elapsed < minInterval {
            try? await Task.sleep(for: .seconds(minInterval - elapsed))
        }
        lastRequestStart = Date()
    }
}

struct NIMClient {
    var model: String

    struct Message: Encodable {
        let role: String
        let content: String
    }

    /// Sends a chat completion to the active provider (Settings → Account)
    /// and returns the assistant's message content. The API key comes from
    /// the in-memory secrets cache (one Keychain read per launch).
    func complete(messages: [Message], maxTokens: Int = 800, temperature: Double = 0.1) async throws -> String {
        let provider = Preferences.shared.provider
        guard let apiKey = Secrets.shared.get(provider.secretKey), !apiKey.isEmpty else {
            throw NIMError.missingAPIKey
        }
        guard let endpoint = Preferences.shared.effectiveBaseURL else {
            throw NIMError.http(-1, "No endpoint URL configured for the custom provider")
        }

        await RequestThrottle.shared.waitTurn()

        var request = URLRequest(url: endpoint)
        request.httpMethod = "POST"
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let payload = RequestBody(
            model: model,
            messages: messages,
            temperature: temperature,
            max_tokens: maxTokens,
            // NVIDIA-only knob (disables reasoning traces for clean JSON);
            // other providers can reject unknown fields, so omit it there.
            chat_template_kwargs: provider == .nvidia ? .init(enable_thinking: false) : nil
        )
        request.httpBody = try JSONEncoder().encode(payload)

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw NIMError.emptyResponse }
        if http.statusCode == 429 { throw NIMError.rateLimited }
        guard (200..<300).contains(http.statusCode) else {
            throw NIMError.http(http.statusCode, String(data: data, encoding: .utf8) ?? "")
        }

        let decoded = try JSONDecoder().decode(ResponseBody.self, from: data)
        guard let content = decoded.choices.first?.message.content, !content.isEmpty else {
            throw NIMError.emptyResponse
        }
        return content
    }

    // MARK: - Wire types

    private struct RequestBody: Encodable {
        struct ChatTemplateKwargs: Encodable { let enable_thinking: Bool }
        let model: String
        let messages: [Message]
        let temperature: Double
        let max_tokens: Int
        let chat_template_kwargs: ChatTemplateKwargs?
    }

    private struct ResponseBody: Decodable {
        struct Choice: Decodable {
            struct Msg: Decodable { let content: String? }
            let message: Msg
        }
        let choices: [Choice]
    }
}

File: JobTracker/Services/NimClient.cs
// Client for the active LLM provider's OpenAI-compatible chat completions
// endpoint. The API key is read from Secrets at call time.

using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace JobTracker.Services;

public enum NimErrorKind { MissingApiKey, Http, RateLimited, EmptyResponse }

public sealed class NimException(NimErrorKind kind, string? detail = null) : Exception(Describe(kind, detail))
{
    public NimErrorKind Kind { get; } = kind;

    private static string Describe(NimErrorKind kind, string? detail) => kind switch
    {
        NimErrorKind.MissingApiKey => "No API key set. Add it in Settings.",
        NimErrorKind.Http => $"LLM API error: {detail}",
        NimErrorKind.RateLimited => "LLM API rate limit hit; will retry.",
        NimErrorKind.EmptyResponse => "LLM API returned no content.",
        _ => "LLM API error",
    };
}

/// Paces all LLM requests to stay under the provider's rate limit (NVIDIA
/// free tier: 40 requests/minute → one request per 1.5s; a small margin
/// keeps bursts safely below the cap). Every request in the app funnels
/// through NimClient, so this is the single enforcement point.
///
/// Relies on callers awaiting one request at a time (true throughout this
/// app — SyncPipeline classifies batches sequentially), matching the
/// macOS version's @MainActor-serialized equivalent.
public sealed class RequestThrottle
{
    public static readonly RequestThrottle Shared = new();
    private static readonly TimeSpan MinInterval = TimeSpan.FromSeconds(1.6);
    private DateTimeOffset _lastRequestStart = DateTimeOffset.MinValue;

    private RequestThrottle() { }

    /// Waits until it's safe to start the next request.
    public async Task WaitTurnAsync()
    {
        var elapsed = DateTimeOffset.UtcNow - _lastRequestStart;
        if (elapsed < MinInterval)
        {
            await Task.Delay(MinInterval - elapsed);
        }
        _lastRequestStart = DateTimeOffset.UtcNow;
    }
}

public sealed class NimClient(string model)
{
    private static readonly HttpClient Http = new();

    public sealed record Message(string Role, string Content);

    /// Sends a chat completion to the active provider (Settings → Account)
    /// and returns the assistant's message content.
    public async Task<string> CompleteAsync(List<Message> messages, int maxTokens = 800, double temperature = 0.1)
    {
        var provider = Preferences.Shared.Provider;

        // Bedrock isn't an OpenAI-compatible HTTP endpoint (AWS SigV4 auth,
        // Converse API request/response shape) — hand off entirely.
        if (provider == LlmProvider.Bedrock)
        {
            return await BedrockClient.CompleteAsync(model, messages, maxTokens, temperature);
        }

        var apiKey = Secrets.Shared.Get(provider.SecretKey());
        if (string.IsNullOrEmpty(apiKey)) throw new NimException(NimErrorKind.MissingApiKey);
        var endpoint = Preferences.Shared.EffectiveBaseUrl
            ?? throw new NimException(NimErrorKind.Http, "No endpoint URL configured for the custom provider");

        await RequestThrottle.Shared.WaitTurnAsync();

        var payload = new RequestBody
        {
            Model = model,
            Messages = messages.Select(m => new WireMessage { Role = m.Role, Content = m.Content }).ToList(),
            Temperature = temperature,
            MaxTokens = maxTokens,
            // NVIDIA-only knob (disables reasoning traces for clean JSON);
            // other providers can reject unknown fields, so omit it there.
            ChatTemplateKwargs = provider == LlmProvider.Nvidia ? new ChatTemplateKwargs { EnableThinking = false } : null,
        };

        using var request = new HttpRequestMessage(HttpMethod.Post, endpoint);
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", apiKey);
        request.Content = new StringContent(JsonSerializer.Serialize(payload), Encoding.UTF8, "application/json");

        using var response = await Http.SendAsync(request);
        var body = await response.Content.ReadAsStringAsync();

        if (response.StatusCode == HttpStatusCode.TooManyRequests) throw new NimException(NimErrorKind.RateLimited);
        if (!response.IsSuccessStatusCode) throw new NimException(NimErrorKind.Http, $"{(int)response.StatusCode}: {body}");

        var decoded = JsonSerializer.Deserialize<ResponseBody>(body);
        var content = decoded?.Choices?.FirstOrDefault()?.Message?.Content;
        if (string.IsNullOrEmpty(content)) throw new NimException(NimErrorKind.EmptyResponse);
        return content;
    }

    // MARK: - Wire types

    private sealed class WireMessage
    {
        [JsonPropertyName("role")] public string Role { get; set; } = "";
        [JsonPropertyName("content")] public string Content { get; set; } = "";
    }

    private sealed class ChatTemplateKwargs
    {
        [JsonPropertyName("enable_thinking")] public bool EnableThinking { get; set; }
    }

    private sealed class RequestBody
    {
        [JsonPropertyName("model")] public string Model { get; set; } = "";
        [JsonPropertyName("messages")] public List<WireMessage> Messages { get; set; } = [];
        [JsonPropertyName("temperature")] public double Temperature { get; set; }
        [JsonPropertyName("max_tokens")] public int MaxTokens { get; set; }
        [JsonPropertyName("chat_template_kwargs")] public ChatTemplateKwargs? ChatTemplateKwargs { get; set; }
    }

    private sealed class ResponseBody
    {
        [JsonPropertyName("choices")] public List<Choice>? Choices { get; set; }
    }

    private sealed class Choice
    {
        [JsonPropertyName("message")] public ChoiceMessage? Message { get; set; }
    }

    private sealed class ChoiceMessage
    {
        [JsonPropertyName("content")] public string? Content { get; set; }
    }
}
